### ISSUE TITLE
False positive on list of union

### ISSUE THREAD

**Issue by PowerSnail** [microsoft/pyright#2347]
**Describe the bug**

A false positive when the function has a parameter of the type is a list of a union. Curiously this only happens if the passed argument is a variable. If the expression is constructed in place, there's no error reported.

**To Reproduce**

See code below.

**Expected behavior**

No error.

**Screenshots or Code**

```python
import typing

int_or_float = typing.Union[int, float]

def take_number(p: int_or_float): ...
def take_number_list(p: list[int_or_float]): ...

n = 1
take_number(n)  # Without a list: No Error

take_number_list([n])  # With a list constructed in line: No Error

n_list = [n]
take_number_list(n_list)  # With a variable: Error
```

I'm getting the error:

```
Argument of type "list[int]" cannot be assigned to parameter "p" of type "list[int_or_float]" in function "number_f2"
    TypeVar "_T@list" is invariant
      Type "int" cannot be assigned to type "int_or_float"
        "int" is incompatible with "float" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**

CLI.

**Comment by erictraut** [comment]
The current behavior is correct. This isn't a false positive error.

Please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md) for details.

If you still have questions after reading the docs, let us know, and we can provide further clarification.
